# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):
    return x + 2

print((lambda x: x+2)(4))

6


#### Saving to a Variable

In [2]:
f_test = lambda x: x + 2
f_test(6)


8

#### Multiple Inputs

In [3]:
#multiple inputs

print((lambda x, y, z: x * y * z)(3, 5, 8))

x_test = lambda x, y, z: x * y * z
print(x_test(3,5,8)) 

120
120


#### Passing a Lambda into a Function

In [4]:
def multiply(f, num):
    return f(num)

multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [5]:
#REGULAR

def multiply_test(num):
    return num * 4

def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

def returnLamb(b,c):
    print(b,c)
    return lambda x, a: x + a + b + c

r_lamb = returnLamb(4,5)
print(r_lamb(6,7))





<function returnFunc.<locals>.multiply at 0x1130e16c0>
8
4 5
22


In [6]:
def addStuff(num1, num2):
    print(num1, num2)
    return lambda x, y: x + y + num1 + num2



lambdaFunc = addStuff(10,20)
print(lambdaFunc(30,40))

10 20
100


#### If Statements within Lambdas

In [7]:
f_condition = lambda num: num * 2 if num > 10 else num + 2

print(f_condition(8))
print(f_condition(12))

10
24


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [8]:
print((lambda x: x**2)(6))



f = lambda x: x**2

print(f(7))



36
49


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [9]:
def squared(num, num2):
    return num*num2

numbers = [4,11, 20, 15, 7, 8]
more_numbs = [5, 7, 2, 10]

squaredNumsMap = list(map(squared, numbers, more_numbs))

print(squaredNumsMap)

[20, 77, 40, 150]


#### Using Lambda's with Map

In [10]:
#map(lambda x: x+2, list_1)
#using lambda in map happpens in one line!
#list(map(lambda x: x*2, nums))

squared_nums_lamb = list(map(lambda x, y: (x ** 2, y **2) if x < 10 and y <10 else (x,y), numbers, more_numbs))

print(squared_nums_lamb)


[(16, 25), (11, 7), (20, 2), (15, 10)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [11]:
numbers = [4,11, 20, 15, 7, 8]

m=list(map(lambda x: x*2-1 if x<10 else x, numbers))
print(m)

l=[x*2-1 if x<10 else x for x in numbers]
print(l)


[7, 11, 20, 15, 13, 15]
[7, 11, 20, 15, 13, 15]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [12]:
names = ['alex', 'beck', 'cecilia', 'dylan', 'julia']

def end_a(name):
    if name[-1].lower() == "a":
        return True
    else: 
        return False
    
    
print(list(filter(end_a, names)))



['cecilia', 'julia']


#### Using Lambda's with Filter()

In [13]:
new_names_lamb = list(filter(lambda name: True if name[-1] == "a" else False, names))

print(new_names_lamb)

['cecilia', 'julia']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [14]:
from statistics import mean

lst = [4, 24, 6, 37, 13, 11, 5]
print(list(filter(lambda x: True if x < mean(lst) else False, lst)))

[4, 6, 13, 11, 5]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [15]:
from functools import reduce

#reduce(function, iterable)

list1 = [2, 4, 6, 8]

def addNums(x, y):
    return x + y

result = reduce(addNums, list1)

print(result)

def subtractNums(x, y):
    return x-y

print(reduce(subtractNums, list1))

print(reduce(addNums, range(100)))

20
-16
4950


#### Using Lambda's with Reduce()

In [16]:
from functools import reduce

reduce_lamb = reduce(lambda x,y: x+y, list1)

print(reduce_lamb)

20


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [17]:
my_list = [3, 6, 5, 4]

print(reduce(lambda x, y: x*y, my_list))


360


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [18]:
def addNums(num):
    if num <= 1:
        print("addNums(1)=1")
        return num
    else:
        print(num)
        return num + addNums(num-1)
    
addNums(5)

5
4
3
2
addNums(1)=1


15

#### Writing a Factorial Function

In [19]:
def factorial(num):
    if num <= 1:
        return 1
    else:
        print(num, num-1, num * num-1)
        return num * factorial(num-1)
    
print(factorial(7))

7 6 48
6 5 35
5 4 24
4 3 15
3 2 8
2 1 3
5040


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [20]:
def my_func(num):
    if num <= 1:
        return 1
    else:
        print(num, num-1)
        return num - my_func(num-1)
    
print(my_func(12))

12 11
11 10
10 9
9 8
8 7
7 6
6 5
5 4
4 3
3 2
2 1
6


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [21]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start 
        start += step
    
gen_object = my_range(20, 2)
for i in gen_object:
    my_generator_value = i
    print(my_generator_value)

print(gen_object)

2
4
6
8
10
12
14
16
18
<generator object my_range at 0x113b73a70>


#### Infinite Generator

In [22]:
# bad, never create infinite loops

#don't forgrt this part!! start += step


#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [23]:
def new_range(stop, start=0, step = 1):
    while start < stop:
        yield stop**2
        stop -= step
        

new_object = new_range(7)
for i in new_object:
    print(i)
    
print(new_object)






49
36
25
16
9
4
1
<generator object new_range at 0x113b73ae0>


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [24]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]


print(list(filter(lambda name: True if name.strip() != "" else False, places)))





['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [25]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]


print(sorted(author, key=lambda x: x.split(' ')[-1].lower()))


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [26]:
# F = (9/5)*C + 32
place = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

# #BRUTE FORCE!

# def conv_func(arr):
#     new_list = []
#     for x in arr:
#         x = list(x)
#         x[1] = (9/5)*(x[1]) + 32
#         new_list.append(tuple(x))
#     return new_list

# print(conv_func(places))

print(list(map(lambda x: (x[0], ((9/5)*x[1] + 32)), place)))




[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [27]:
#This is not right :( I can't get it to work. 
def fibonacci(x):
    if x <= 0:
        return x
    else:
        print(f"Iteration {x}: {x+x}")
        return x + fibonacci(x-1)

fibonacci(5)

#This also does not work. 
# def fibonacci2(x):
#     a=0
#     b=1
#     c=0
#     while a <= x:
#         print(f"Iteration {a}: {b}")
#         a += 1
#         b = b+c
#         c=b
      
# fibonacci2(5)
           

Iteration 5: 10
Iteration 4: 8
Iteration 3: 6
Iteration 2: 4
Iteration 1: 2


15